File: 2022/Javascript/src/Day11.js
/**
 * An object representing a Monkey from the input data.
 */
class Monkey {
	#id
	#op_mul
	#op_const
	#op_sec
	#rel_mod
	#rel_div
	#test_div
	#true_tgt
	#false_tgt
	#items
	#inspected

	/**
	 * Creates a new monkey with the given properties.
	 *
	 * @param id                 The numeric id of this monkey.
	 * @param start_items        The items this monkey holds when it is first created.
	 * @param operation_multiply Whether this monkey should multiply the items value, instead of adding to it.
	 * @param operation_const    Whether the second value for the item value modification is a constant.
	 * @param operation_second   The second value for the operation. Ignored if operation_const is false.
	 * @param relief_mod         Whether the worry reduction should be a modulo operation instead of a division.
	 * @param relief_divisor     The divisor for the worry reduction operation.
	 * @param test_divisor       The divisor used to check which monkey to throw items to.
	 * @param true_target        The monkey to throw items for which the test returns true to.
	 * @param false_target       The monkey to throw items for which the test returns false to.
	 */
	constructor(id, start_items, operation_multiply, operation_const, operation_second, relief_mod,
		relief_divisor, test_divisor, true_target, false_target) {
		this.#id = id
		this.#op_mul = operation_multiply
		this.#op_const = operation_const
		this.#op_sec = operation_second
		this.#rel_mod = relief_mod
		this.#rel_div = relief_divisor
		this.#test_div = test_divisor
		this.#true_tgt = true_target
		this.#false_tgt = false_target
		this.#items = start_items
		this.#inspected = 0
	}

	/**
	 * Executes a single turn for this monkey.
	 */
	turn() {
		for (let i in this.#items) {
			let second = 1
			if (this.#op_const) {
				second = this.#op_sec
			} else {
				second = this.#items[i]
			}

			if (this.#op_mul) {
				this.#items[i] *= second
			} else {
				this.#items[i] += second
			}

			this.#inspected++

			if (this.#rel_mod) {
				this.#items[i] %= this.#rel_div
			} else {
				this.#items[i] = Math.floor(this.#items[i] / this.#rel_div)
			}

			if (this.#items[i] % this.#test_div == 0) {
				this.#true_tgt.#items.push(this.#items[i])
			} else {
				this.#false_tgt.#items.push(this.#items[i])
			}
		}
		this.#items = []
	}

	/**
	 * Set the targets to throw items to.
	 *
	 * @param true_target  The target monkey for items for which the test succeeds.
	 * @param false_target The target monkey for items for which the test fails.
	 */
	set_targets(true_target, false_target) {
		if (true_target != undefined && !(true_target instanceof Monkey)) {
			throw new TypeError("true_target has to be a Monkey")
		}
		if (false_target != undefined && !(false_target instanceof Monkey)) {
			throw new TypeError("false_target has to be a Monkey")
		}

		this.#true_tgt = true_target
		this.#false_tgt = false_target
	}

	/**
	 * The number of items this monkey has already inspected.
	 */
	get inspected() {
		return this.#inspected
	}

	/**
	 * The divisor to use to determine which monkey an item should be thrown to.
	 */
	get test_divisor() {
		return this.#test_div
	}

	/**
	 * Sets how the worry value should be reduced after a monkey inspected an item.
	 *
	 * @param relief_mod     Whether the worry reduction should be a modulo operation instead of a division.
	 * @param relief_divisor The divisor for the worry reduction operation.
	 */
	set_relief(relief_mod, relief_divisor) {
		this.#rel_mod = relief_mod
		this.#rel_div = relief_divisor
	}

	/**
	 * Converts this monkey to its string representation.
	 * 
	 * @return The string representation of this monkey.
	 */
	toString() {
		return `Monkey ${this.#id}: ${this.#items.join(", ")}`
	}
}

/**
 * Parses the monkeys from the input file as objects.
 *
 * @param lines The input file lines to parse.
 * @return A list containing the objects representing the monkeys.
 */
function parseMonkeys(lines) {
	var monkeys = []
	var targets = []
	var id
	var items = []
	var op_mul
	var op_const
	var op_sec
	var test_div
	// Which of the 5 lines of a monkey definition are we parsing
	var ln = 0;
	for (let i in lines) {
		if (lines[i].length == 0) {
			continue
		}

		lines[i] = lines[i].trim()

		let err = false
		switch (ln) {
			case 0:
				if (!lines[i].startsWith("Monkey")) {
					console.error(`Trying to parse invalid id line "${lines[i]}.`)
					err = true
				} else {
					id = Number(lines[i].substring(7, lines[i].length - 1))
				}
				break
			case 1:
				if (!lines[i].startsWith("Starting items")) {
					console.error(`Trying to parse invalid items line "${lines[i]}".`)
					err = true
				} else {
					items = []
					let split = lines[i].substring(16).split(", ")
					for (let j in split) {
						items.push(Number(split[j]))
					}
				}
				break
			case 2:
				if (!lines[i].startsWith("Operation")) {
					console.error(`Trying to parse invalid operation line "${lines[i]}.`)
					err = true
				} else {
					op_mul = lines[i][21] == '*'
					let nr = lines[i].substr(23)
					op_const = nr != "old"
					op_sec = op_const ? Number(nr) : 0
				}
				break
			case 3:
				if (!lines[i].startsWith("Test")) {
					console.error(`Trying to parse invalid test line "${lines[i]}.`)
					err = true
				} else {
					test_div = Number(lines[i].substring(19))
				}
				break
			case 4:
				if (!lines[i].startsWith("If true")) {
					console.error(`Trying to parse invalid if true line "${lines[i]}.`)
					err = true
				} else {
					targets.push({ true: Number(lines[i].substring(25)) })
				}
				break
			case 5:
				if (!lines[i].startsWith("If false")) {
					console.error(`Trying to parse invalid if false line "${lines[i]}.`)
					err = true
				} else {
					targets[targets.length - 1].false = Number(lines[i].substring(26))
				}
				break;
		}

		if (err) {
			ln = 0
		} else if (ln == 5) {
			monkeys.push(new Monkey(
				id, items, op_mul, op_const, op_sec, false, 3, test_div, undefined, undefined))
			ln = 0
		} else {
			ln++
		}
	}

	for (let i in monkeys) {
		monkeys[i].set_targets(monkeys[targets[i].true], monkeys[targets[i].false])
	}

	return monkeys
}

function main() {
	const path = require('path')
	const main = require(path.join(path.resolve(''), 'Main.js'))
	const os = require('os')
	const lines = main.getDayInput(11).split(os.EOL)

	var monkeys = parseMonkeys(lines)
	for (let i = 0; i < 20; i++) {
		for (let j in monkeys) {
			monkeys[j].turn()
		}
	}

	var inspections = []
	for (let i in monkeys) {
		inspections.push(monkeys[i].inspected)
	}
	inspections.sort(function(a, b) { return b - a })

	console.log(`The part 1 monkey business is ${inspections[0] * inspections[1]}.`)

	monkeys = parseMonkeys(lines)
	var lcm = monkeys[0].test_divisor;
	var gcd
	var a
	var b
	for (let i in monkeys) {
		gcd = monkeys[i].test_divisor
		a = lcm
		while (a != 0) {
			b = a
			a = gcd % a
			gcd = b
		}
		lcm = (monkeys[i].test_divisor * lcm) / gcd;
	}

	for (let i in monkeys) {
		monkeys[i].set_relief(true, lcm)
	}

	for (let i = 0; i < 10000; i++) {
		for (let j in monkeys) {
			monkeys[j].turn()
		}
	}

	inspections = []
	for (let i in monkeys) {
		inspections.push(monkeys[i].inspected)
	}
	inspections.sort(function(a, b) { return b - a })

	console.log(`The part 2 monkey business is ${inspections[0] * inspections[1]}.`)
}

main()

File: 2022/Cpp/src/Day11.cpp
/*
 * Day11.cpp
 *
 *  Created on: Dec 11, 2022
 *      Author: ToMe25
 */

#include "Day11.h"
#include <algorithm>
#include <cctype>

aoc::Monkey::Monkey(const uint8_t id, const std::vector<uint64_t> start_items,
		const bool operation_multiply, const bool operation_const,
		const uint8_t operation_second, const bool relief_mod,
		const uint32_t relief_divisor, const uint8_t test_divisor,
		Monkey *true_target, Monkey *false_target) :
		id(id), op_mul(operation_multiply), op_const(operation_const), op_sec(
				operation_second), rel_mod(relief_mod), rel_div(relief_divisor), test_div(
				test_divisor), true_tgt(true_target), false_tgt(false_target), items(
				start_items) {
}

aoc::Monkey::Monkey(const Monkey &monkey) :
		id(monkey.id), op_mul(monkey.op_mul), op_const(monkey.op_const), op_sec(
				monkey.op_sec), rel_mod(monkey.rel_mod), rel_div(
				monkey.rel_div), test_div(monkey.test_div), true_tgt(
				monkey.true_tgt), false_tgt(monkey.false_tgt), items(
				monkey.items) {
}

void aoc::Monkey::set_targets(Monkey *true_target, Monkey *false_target) {
	true_tgt = true_target;
	false_tgt = false_target;
}

void aoc::Monkey::set_relief(const bool relief_mod,
		const uint32_t relief_divisor) {
	rel_mod = relief_mod;
	rel_div = relief_divisor;
}

uint64_t aoc::Monkey::get_inspected() const {
	return inspected;
}

uint32_t aoc::Monkey::get_test_divisor() const {
	return test_div;
}

void aoc::Monkey::turn() {
	for (size_t i = 0; i < items.size(); i++) {
		if (op_mul) {
			items[i] *= (op_const ? op_sec : items[i]);
		} else {
			items[i] += (op_const ? op_sec : items[i]);
		}
		inspected++;

		if (rel_mod) {
			items[i] %= rel_div;
		} else {
			items[i] /= rel_div;
		}

		if (items[i] % test_div == 0) {
			if (true_tgt == NULL) {
				std::cerr << "Monkey " << (uint16_t) id
						<< " tried to throw an item to a null monkey."
						<< std::endl;
			} else {
				true_tgt->items.push_back(items[i]);
			}
		} else {
			if (false_tgt == NULL) {
				std::cerr << "Monkey " << (uint16_t) id
						<< " tried to throw an item to a null monkey."
						<< std::endl;
			} else {
				false_tgt->items.push_back(items[i]);
			}
		}
	}
	items.clear();
}

std::vector<aoc::Monkey> aoc::parse_monkeys(std::vector<std::string> lines) {
	std::vector<aoc::Monkey> monkeys;
	std::vector<std::pair<uint8_t, uint8_t>> targets;
	uint8_t id = 0;
	std::vector<uint64_t> items;
	bool operation_multiply = false;
	// -1 to use old value.
	int8_t operation_second = 0;
	uint8_t test_divisor = 0;
	//Which of the 5 lines of a monkey definition are we parsing
	uint8_t ln = 0;
	for (std::string line : lines) {
		if (line.length() == 0) {
			continue;
		}

		bool err = false;
		switch (ln) {
		case 0:
			if (line.substr(0, 6) != "Monkey") {
				std::cerr << "Trying to parse invalid id line \"" << line
						<< "\"." << std::endl;
				err = true;
			} else {
				id = std::stoi(line.substr(7, line.length() - 8));
			}
			break;
		case 1:
			if (line.substr(0, 14) != "Starting items") {
				std::cerr << "Trying to parse invalid items line \"" << line
						<< "\"." << std::endl;
				err = true;
			} else {
				items = std::vector<uint64_t>();
				line = line.substr(16);
				size_t comma;
				while ((comma = line.find(',')) != std::string::npos) {
					items.push_back(std::stoi(line.substr(0, comma)));
					line = line.substr(comma + 2);
				}
				items.push_back(std::stoi(line));
			}
			break;
		case 2:
			if (line.substr(0, 9) != "Operation") {
				std::cerr << "Trying to parse invalid operation line \"" << line
						<< "\"." << std::endl;
				err = true;
			} else {
				operation_multiply = line[21] == '*';
				line = line.substr(23);
				if (line == "old") {
					operation_second = -1;
				} else {
					operation_second = std::stoi(line);
				}
			}
			break;
		case 3:
			if (line.substr(0, 4) != "Test") {
				std::cerr << "Trying to parse invalid test line \"" << line
						<< "\"." << std::endl;
				err = true;
			} else {
				test_divisor = std::stoi(line.substr(19));
			}
			break;
		case 4:
			if (line.substr(0, 7) != "If true") {
				std::cerr << "Trying to parse invalid if true line \"" << line
						<< "\"." << std::endl;
				err = true;
			} else {
				targets.push_back( { std::stoi(line.substr(25)), 0 });
			}
			break;
		case 5:
			if (line.substr(0, 8) != "If false") {
				std::cerr << "Trying to parse invalid if false line \"" << line
						<< "\"." << std::endl;
				err = true;
			} else {
				targets.back().second = std::stoi(line.substr(26));
			}
			break;
		}

		if (err) {
			ln = 0;
		} else if (ln == 5) {
			monkeys.push_back(
					aoc::Monkey(id, items, operation_multiply,
							operation_second != -1,
							operation_second == -1 ? 0 : operation_second,
							false, 3, test_divisor));
			ln = 0;
		} else {
			ln++;
		}
	}

	for (size_t i = 0; i < monkeys.size(); i++) {
		monkeys[i].set_targets(&monkeys[targets[i].first],
				&monkeys[targets[i].second]);
	}

	return monkeys;
}

std::ostream& aoc::operator<<(std::ostream &stream, const aoc::Monkey &monkey) {
	stream << "Monkey " << (uint16_t) monkey.id << ": ";
	for (size_t i = 0; i < monkey.items.size(); i++) {
		stream << monkey.items[i];
		if (i < monkey.items.size() - 1) {
			stream << ", ";
		}
	}
	return stream;
}

std::string day11part1(std::ifstream input) {
	std::vector<std::string> lines;
	std::string line;
	while (std::getline(input, line)) {
		if (line.length() > 0) {
			line.erase(line.begin(),
					std::find_if_not(line.begin(), line.end(),
							(int (*)(int)) (std::isspace)));
			lines.push_back(line);
		}
	}
	std::vector<aoc::Monkey> monkeys = aoc::parse_monkeys(lines);

	for (uint8_t i = 0; i < 20; i++) {
		for (aoc::Monkey &monkey : monkeys) {
			monkey.turn();
		}
	}

	std::vector<uint32_t> inspections;
	for (const aoc::Monkey &m : monkeys) {
		inspections.push_back(m.get_inspected());
	}
	std::sort(inspections.begin(), inspections.end(), std::greater());

	return std::to_string(inspections[0] * inspections[1]);
}

std::string day11part2(std::ifstream input) {
	std::vector<std::string> lines;
	std::string line;
	while (std::getline(input, line)) {
		if (line.length() > 0) {
			line.erase(line.begin(),
					std::find_if_not(line.begin(), line.end(),
							(int (*)(int)) (std::isspace)));
			lines.push_back(line);
		}
	}
	std::vector<aoc::Monkey> monkeys = aoc::parse_monkeys(lines);

	uint64_t lcm = monkeys[0].get_test_divisor();
	uint64_t gcd, a, b;
	for (const aoc::Monkey &monkey : monkeys) {
		gcd = monkey.get_test_divisor();
		a = lcm;
		while (a != 0) {
			b = a;
			a = gcd % a;
			gcd = b;
		}
		lcm = (monkey.get_test_divisor() * lcm) / gcd;
	}

	for (aoc::Monkey &m : monkeys) {
		m.set_relief(true, lcm);
	}

	for (uint16_t i = 0; i < 10000; i++) {
		for (aoc::Monkey &monkey : monkeys) {
			monkey.turn();
		}
	}

	std::vector<uint64_t> inspections;
	for (const aoc::Monkey &m : monkeys) {
		inspections.push_back(m.get_inspected());
	}
	std::sort(inspections.begin(), inspections.end(), std::greater());

	return std::to_string(inspections[0] * inspections[1]);
}

bool d11p1 = aoc::registerPart1(11, &day11part1);
bool d11p2 = aoc::registerPart2(11, &day11part2);
